### PR TITLE
fix(common): fix wrong memcomparable encoding for struct/list/interval

### PR DESCRIPTION
**Pull request by wzzzzd**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Now we can compare serialized values of struct/list/interval correctly.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Comment by xiangjinwu** [comment]
With this encoding change I am able to produce a panic:
```
create table t (v1 interval);
insert into t values (interval '24 h'), (interval '1 d');
create materialized view mv1 as select distinct v1 from t;
```

The error message is:
```
thread 'tokio-runtime-worker' panicked at 'invalid flush status: double insert [0, 128, 0, 0, 0, 128, 0, 0, 1, 128, 0, 0, 0, 0, 0, 0, 0] -> Row([Some(Interval(IntervalUnit { months: 0, days: 0, ms: 86400000 })), Some(Int64(1))])', src/storage/src/table/mem_table.rs:75:21
```

I did not trace the error in streaming distinct. But my hypothesis was that `streaming distinct` produced 2 distinct values while the storage layer was unhappy because these 2 keys share the same memcomparable encoding, thus "double insert". The behavior of producing 2 distinct values can be verified in batch #3873

It can be "fixed" into the same behavior as decimal, to output a single normalized/justified value, at this place:
https://github.com/singularity-data/risingwave/blob/8135a39585d6f5bfe9d58697165a9a3b1fd9ea1a/src/common/src/hash/key.rs#L317-L340

Note that decimal is `normalize`d then `unordered_serialize`, but interval is encoded as 3 parts without `justify`.

To summarize:

`HashKeySerDe` seems to be neither memcomparable encoding nor value encoding, and is potentially inconsistent with `Hash` and `Eq` defined the element type. Do we have other serializations (https://github.com/singularity-data/risingwave/issues/13#issuecomment-1072300164)?

Related issues:
* #3873
* #3863

This might be out of the scope for this PR. If we cannot conclude on a solution easily, my suggestion is to fix the panic via `justify` in `HashKeySerDe` here, and discuss proper handling of distinct / pk / hash for types incl. interval / decimal / real / double separately.

**Comment by liurenjie1024** [comment]
After offline discussion, I think @xiangjinwu 's suggestion is great. And this leads to the behavior of pg. 😂

**Review by liurenjie1024 — APPROVED** [review]
LGTM
